TypeScript_ESLint.snb.md

```diff
@@ -2,92 +2,176 @@
+- `messageId`: Which string ID of message to report, from `meta`
+- `fix`: A function that generate instructions for how to auto-fix the code issue
+
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/rules/no-extra-non-null-assertion.ts?L20-30
 
 ## Type-checking
 
-One distinction between TypeScript ESLint and vanilla ESLint is that we have access to the TypeScript type checker and can therefore define rules that make use of type information.
+One powerful distinction between TypeScript ESLint and vanilla ESLint is that we have access to the TypeScript type checker.
+Rules that make use of type checking are able to make informed decisions on nodes based on their type information.
+
+TypeScript ESLint's `no-for-in-array` rule uses type information to determine if a `for-in` loop is iterating over an array rather than an object or string.
+Knowing the type of a value is only possible in JavaScript or TypeScript code if you have access to a type checker.
+Here is an example of this pattern:
+
+https://sourcegraph.com/github.com/cypress-io/cypress/-/blob/packages/driver/src/cypress/proxy-logging.ts?L13
+
+This anti-pattern is quite widespread, because it feels intuitive to use the for-in syntax to traverse an array.
+Doing so, however, may visit the array elements out-of-order.
+Instead `for-of` loops or `array.forEach` are recommended.
+
+ESLint rules that use type information are set up similarly to other ESLint rules.
+Before we jump into how they use the type checker, the source of `no-for-in-array` is also small enough to be seen in its entirety here:
+
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/rules/no-for-in-array.ts
+
+## Getting a Type Checker
 
-The `for-in-array` rule uses the type checker to determine if we are dealing with an array type:
+The `no-for-in-array` rule enables a `meta.docs.requiresTypeChecking` property indicating it plans on using type checking powered by a TypeScript type checker.
 
-https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/rules/no-for-in-array.ts?L23-24
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/rules/no-for-in-array.ts?10
 
-https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/rules/no-for-in-array.ts?L32-35
+A TypeScript type checker may be retrieved by a rule during its runtime by retrieving TypeScript ESLint's "parser services" for its `getTypeChecker()` API:
 
-## Dual ASTs
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/rules/no-for-in-array.ts?23-24
 
-In addition to invoking the type checker, the `for-in-array` rule also makes use of **two ASTs**:
+### Dual ASTs
```

Comment:
@beyang this section is much more in-depth than the others. The concepts of recursive AST creation/mapping and dual ASTs are very cool but are already covered at a high level by the earlier sections of the blog post. What do you think of removing this deeper explanation altogether from the blog post, and saving it for the TypeScript ESLint docs?

I added a bunch of content earlier that goes line-by-line, so it feels like this extra level detail is no longer really necessary.

Reply:
Removed `### ASTs Behind The Scenes` in b04fa13d